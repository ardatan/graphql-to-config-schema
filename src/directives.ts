export const DIRECTIVES = /* GraphQL */ `
  directive @md on OBJECT
`;
