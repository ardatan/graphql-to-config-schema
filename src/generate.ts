import {
  GraphQLSchema,
  printSchema,
  parse,
  visit,
  FieldDefinitionNode,
  Kind,
  TypeNode,
  NamedTypeNode,
  isScalarType,
  isEnumType,
  isObjectType,
  isInterfaceType,
  GraphQLInterfaceType,
  GraphQLObjectType,
  isUnionType
} from 'graphql';
import { JSONSchema7, JSONSchema7Definition } from 'json-schema';

export function generateFromSchema(
  schema: GraphQLSchema,
  rootType = 'Query'
): JSONSchema7 {
  const documentNode = parse(printSchema(schema));
  const jsonSchema: JSONSchema7 = {
    definitions: {},
    title: 'Config',
    type: 'object',
    $schema: 'http://json-schema.org/draft-07/schema#'
  };

  visit(documentNode, {
    ObjectTypeDefinition(node) {
      if (node.name.value === rootType) {
        (jsonSchema.required = node.fields
          .filter(f => f.type.kind === Kind.NON_NULL_TYPE)
          .map(f => f.name.value)),
          (jsonSchema.properties = buildPropertiesFromFields(
            schema,
            node.fields
          ));
        return;
      }

      jsonSchema.definitions[node.name.value] = {
        additionalItems: false,
        additionalProperties: false,
        type: 'object',
        title: node.name.value,
        required: node.fields
          .filter(f => f.type.kind === Kind.NON_NULL_TYPE)
          .map(f => f.name.value),
        properties: buildPropertiesFromFields(schema, node.fields)
      };
    }
  });

  return jsonSchema;
}

function buildPropertiesFromFields(
  schema: GraphQLSchema,
  fields: ReadonlyArray<FieldDefinitionNode>
): Record<string, JSONSchema7Definition> {
  return fields.reduce((prev, field) => {
    const namedType = getBaseTypeNode(field.type);
    const typeToUse = getTypeToUse(schema, namedType.name.value);
    const isArrayField = isArray(field.type);
    const fieldDef = isArrayField
      ? { type: 'array', items: typeToUse }
      : typeToUse;

    return {
      ...prev,
      additionalItems: false,
      additionalProperties: false,
      [field.name.value]: fieldDef
    };
  }, {});
}

function getBaseTypeNode(typeNode: TypeNode): NamedTypeNode {
  if (
    typeNode.kind === Kind.LIST_TYPE ||
    typeNode.kind === Kind.NON_NULL_TYPE
  ) {
    return getBaseTypeNode(typeNode.type);
  }

  return typeNode;
}

function isArray(typeNode: TypeNode): boolean {
  if (typeNode.kind === Kind.NON_NULL_TYPE) {
    return isArray(typeNode.type);
  } else if (typeNode.kind === Kind.LIST_TYPE) {
    return true;
  }

  return false;
}

const SCALARS = {
  String: 'string',
  ID: 'string', // Should be ["string", "number"] ?
  Boolean: 'boolean',
  Float: 'number',
  Int: 'integer',
  JSON: 'object'
};

function getTypeToUse(
  schema: GraphQLSchema,
  typeName: string
): JSONSchema7Definition {
  const schemaType = schema.getType(typeName);

  if (isScalarType(schemaType) && SCALARS[schemaType.name]) {
    return {
      type: SCALARS[schemaType.name]
    };
  } else if (isEnumType(schemaType)) {
    return {
      type: 'string',
      enum: schemaType.getValues().map(v => v.value)
    };
  } else if (isObjectType(schemaType)) {
    return {
      $ref: `#/definitions/${schemaType.name}`
    };
  } else if (isInterfaceType(schemaType)) {
    return {
      anyOf: getImplementingTypes(schema, schemaType).map(t =>
        getTypeToUse(schema, t)
      )
    };
  } else if (isUnionType(schemaType)) {
    return {
      anyOf: schemaType.getTypes().map(t => getTypeToUse(schema, t.name))
    };
  }

  return {
    type: 'object'
  };
}

function getImplementingTypes(
  schema: GraphQLSchema,
  type: GraphQLInterfaceType
): string[] {
  const allTypesMap = schema.getTypeMap();
  const implementingTypes: string[] = [];

  for (const graphqlType of Object.values(allTypesMap)) {
    if (graphqlType instanceof GraphQLObjectType) {
      const allInterfaces = graphqlType.getInterfaces();
      if (allInterfaces.find(int => int.name === type.name)) {
        implementingTypes.push(graphqlType.name);
      }
    }
  }

  return implementingTypes;
}
