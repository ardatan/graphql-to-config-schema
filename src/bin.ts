#!/usr/bin/env node

import * as yargs from 'yargs';
import { loadSchema } from '@graphql-toolkit/core';
import { UrlLoader } from '@graphql-toolkit/url-loader';
import { JsonFileLoader } from '@graphql-toolkit/json-file-loader';
import { GraphQLFileLoader } from '@graphql-toolkit/graphql-file-loader';
import { generateFromSchema } from './generate';
import { writeFileSync } from 'fs';
import { resolve, join } from 'path';
import { compile } from 'json-schema-to-typescript';
import { DIRECTIVES } from './directives';
import { generatedMarkdown } from './md-generator';

async function main() {
  const { schema, typings, json, rootType, markdown } = yargs
    .option('schema', {
      required: true,
      type: 'array'
    })
    .option('json', {
      required: true,
      type: 'string'
    })
    .option('rootType', {
      type: 'string',
      default: 'Query'
    })
    .option('typings', {
      type: 'string'
    })
    .option('markdown', {
      alias: 'md',
      type: 'string'
    }).argv;

  const unifiedSchemea = await loadSchema(
    [...((schema as string[]) || []), DIRECTIVES],
    {
      loaders: [
        new UrlLoader(),
        new JsonFileLoader(),
        new GraphQLFileLoader(),
        new GraphQLFileLoader()
      ]
    }
  );

  const jsonSchema = generateFromSchema(unifiedSchemea, rootType);
  const schemaFilePath = resolve(process.cwd(), `./${json}`);
  writeFileSync(schemaFilePath, JSON.stringify(jsonSchema, null, 2));

  if (typings) {
    const tsTypes = await compile(jsonSchema, 'Config');
    const tsFilePath = resolve(process.cwd(), `./${typings}`);
    writeFileSync(tsFilePath, tsTypes);
  }

  if (markdown) {
    const mdFiles = await generatedMarkdown(unifiedSchemea);
    const mdDir = resolve(process.cwd(), `./${markdown}`);

    for (const result of mdFiles) {
      const filePath = join(mdDir, result.file);
      writeFileSync(filePath, result.content);
    }
  }
}

main();
