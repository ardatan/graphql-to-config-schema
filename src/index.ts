#!/usr/bin/env node

import * as yargs from 'yargs';
import { loadSchema } from '@graphql-toolkit/core';
import { UrlLoader } from '@graphql-toolkit/url-loader';
import { JsonFileLoader } from '@graphql-toolkit/json-file-loader';
import { GraphQLFileLoader } from '@graphql-toolkit/graphql-file-loader';
import { generateFromSchema } from './generate';
import { writeFileSync } from 'fs';
import { resolve } from 'path';
import { compile } from 'json-schema-to-typescript';

async function main() {
  const { schema, typings, json, rootType } = yargs
    .option('schema', {
      required: true,
      type: 'array'
    })
    .option('json', {
      required: true,
      type: 'string'
    })
    .option('rootType', {
      type: 'string',
      default: 'Query'
    })
    .option('typings', {
      type: 'string'
    }).argv;

  const unifiedSchemea = await loadSchema(schema as string[], {
    loaders: [
      new UrlLoader(),
      new JsonFileLoader(),
      new GraphQLFileLoader(),
      new GraphQLFileLoader()
    ]
  });

  const jsonSchema = generateFromSchema(unifiedSchemea, rootType);
  const schemaFilePath = resolve(process.cwd(), `./${json}`);
  writeFileSync(schemaFilePath, JSON.stringify(jsonSchema, null, 2));

  if (typings) {
    const tsTypes = await compile(jsonSchema as any, 'Config');
    const tsFilePath = resolve(process.cwd(), `./${typings}`);
    writeFileSync(tsFilePath, tsTypes);
  }
}

main();
