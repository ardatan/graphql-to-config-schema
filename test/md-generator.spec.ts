import { DIRECTIVES } from '../src/directives';
import { buildSchema } from 'graphql';
import { generatedMarkdown } from '../src/md-generator';

describe('MD Generator', () => {
  it('Should generate md correctly', () => {
    const schema = buildSchema(/*GraphQL*/ `
    ${DIRECTIVES}

    type Query {
      m: MyType
    }

    type MyType @md {
      f1: String
      f2: [String]
      f3: [String]!
      f4: [String!]!
      f5: SubType
      f6: Test
    }

    type SubType {
      t: String
      i: Int!
      nest: Sub2Type!
      other: String!
    }

    enum Test {
      A
      B
      C
    }

    type Sub2Type {
      t: String
    }
    `);

    const mdOutput = generatedMarkdown(schema);
    expect(mdOutput.length).toBe(1);
    expect(mdOutput[0].content).toMatchSnapshot();
  });
});
