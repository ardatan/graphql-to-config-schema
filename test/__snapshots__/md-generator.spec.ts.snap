// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MD Generator Should generate md correctly 1`] = `
" test desc 


* \`f1\` (type: \`String\`) - this
is
multi
* \`f2\` (type: \`Array of String\`)
* \`f3\` (type: \`Array of String\`, required)
* \`f4\` (type: \`Array of String\`, required)
* \`f5\` (type: \`Object\`): 
  * \`t\` (type: \`String\`)
  * \`i\` (type: \`Int\`, required)
  * \`nest\` (type: \`Object\`, required) - test: 
    * \`t\` (type: \`String\`)
  * \`other\` (type: \`String\`, required)
* \`f6\` (type: \`String (A | B | C)\`)
* \`u\` -  One of: 
  * \`Boolean\`
  * \`object\`: 
    * \`boop\` (type: \`String\`)"
`;
